### PR TITLE
*: fix compatibility to `raft-engine.spill-dir`

### LINKED ISSUE #17356: disk-full: TiKV is not compatible to the configuration `spill-dir`.
## Bug Report

In previous work https://github.com/tikv/raft-engine/pull/294, raft-engine introduced the `spill-dir` configuration to enable multi-directory configurations for raft-engine.

However, it has been discovered that this feature is not compatible with TiKV, as the `DiskFull` check does not behave as anticipated and it would ignore the `spill-dir` directory, resulting in the `spill-dir` functionality not working as intended.

### What version of TiKV are you using?

TiKV >= v6.5.0 

### What operating system and CPU are you using?

Linux / MacOS

### Steps to reproduce

1. Set the configuration `raft-engine.spill-dir` manually, and bind the directory to a separate disk with the main `raft-engine.dir`.
2. Start the TiKV, and keeps writing data into it.
3. When the disk of `raft-engine.dir` is almost full, TiKV will reject to append the incoming writes.

### What did you expect?
The new data (raft logs) will be flushed into the `spill-dir` rather than rejecting appends.

### What did happened?
TiKV rejects appends and keeps logging `DiskFull`.
```
[2024/08/05 07:27:49.320 +00:00] [INFO] [[peer.rs:4515](http://peer.rs:4515/)] ["reject transferring leader"] [wait_data=false] [is_witness=false] [disk_usage=AlreadyFull] [pending_snapshot=false] [from=14] [peer_id=3] [region_id=2] [thread_id=82]
```

### PR DESCRIPTION
**Pull request by LykxSassinator**
### What is changed and how it works?

Issue Number: Close https://github.com/tikv/tikv/issues/17356

What's Changed:

```commit-message
Make the diskfull check mechanism compatible to the configuration `raft-engine.spill-dir`.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
Make the diskfull check mechanism compatible to the configuration `raft-engine.spill-dir`.
```

### CODE REVIEW

**Review by SpadeA-Tang — APPROVED** [review]

File: components/server/src/server.rs [hunk 1]
```diff
@@ -1440,7 +1455,7 @@ where
 
                 let mut raft_disk_status = disk::DiskUsage::Normal;
```
Comment (Connor1996) [inline components/server/src/server.rs:1456]: make it follow rust style
```
let raft_disk_status = if {
  ...
} else {
   disk::DiskUsage::Normal
}

**Comment by LykxSassinator** [comment]
Please hold on it, I'll refactor the check strategy as the current processing on checking disk full status is too ambiguous to read and redundant both on `kvdb` and `partition-kv`.

**Comment by LykxSassinator** [comment]
> Please hold on it, I'll tidy the check strategy as the current processing on checking disk full status is too ambiguous to read and redundant both on `kvdb` and `partition-kv`.

Done.

File: components/tikv_util/src/sys/mod.rs [hunk 2]
```diff
@@ -348,4 +350,13 @@ mqueue /dev/mqueue mqueue rw,relatime 0 0
         check("/home", Some("/dev/vda3 / ext4"));
         check("unknown/path", None);
     }
+
+    #[test]
+    fn test_get_disk_space_stats() {
+        let (capacity, available) = disk::get_disk_space_stats("./").unwrap();
+        assert!(capacity > 0);
+        assert!(available > 0);
```
Comment (zhangjinpeng87) [inline components/tikv_util/src/sys/mod.rs:358]: assert!(capacity >= available);

File: components/server/src/common.rs [hunk 3]
```diff
@@ -924,3 +916,338 @@ impl<EK: KvEngine, ER: RaftEngine> EngineMetricsManager<EK, ER> {
...
+    /// Inspect the disk usage of kv engine and raft engine.
+    /// The `used_size` is the used size of kv engine, and the `raft_size` is
+    /// the used size of raft engine.
+    ///
+    /// Returns the disk usage status of the whole disk, kv engine and raft
+    /// engine, the whole disk capacity and available size.
+    pub fn inspect(
+        &self,
+        used_size: u64,
+        raft_size: u64,
```
Comment (zhangjinpeng87) [inline components/server/src/common.rs:993]: s/used_size/db_used_size
s/raft_size/raft_used_size

File: components/server/src/common.rs [hunk 4]
```diff
@@ -924,3 +916,338 @@ impl<EK: KvEngine, ER: RaftEngine> EngineMetricsManager<EK, ER> {
...
+    /// The path of auxiliary directory of raft engine if specified.
+    raft_auxiliary_path: Option<String>,
+    /// Whether the main directory of raft engine is separated from kv engine.
+    separated_raft_mount_path: bool,
+    /// Whether the auxiliary directory of raft engine is separated from kv
+    /// engine.
+    separated_raft_auxiliary_mount_path: bool,
+    /// Whether the auxiliary directory of raft engine is both separated from
+    /// the main directory of raft engine and kv engine.
+    separated_raft_and_kvdb_mount_path: bool,
```
Comment (zhangjinpeng87) [inline components/server/src/common.rs:948]: separated_raft_auxiliary_and_kvdb_mount_path

**Review by zhangjinpeng87 — APPROVED** [review]
rest LGTM

**Pull request merged**